torchaudio/transforms.py

```diff
@@ -284,22 +291,102 @@ class SpectogramToDB(object):
+        spec_db -= self.multiplier * torch.log10(torch.max(self.amin.to(spec.device), self.ref_value.to(spec.device)))
+
         if self.top_db is not None:
-            spec_db = torch.max(spec_db, spec_db.new_full((1,), self.top_db))
+            spec_db = torch.max(spec_db, spec_db.new_full((1,), spec_db.max() - self.top_db))
         return spec_db
 
 
-def MEL2(*args, **kwargs):
-    warn("MEL2 has been renamed to MelSpectrogram")
-    return MelSpectrogram(*args, **kwargs)
+class MFCC(object):
+    """Create the Mel-frequency cepstrum coefficients from an audio signal
+
+        Args:
+        sr (int) : sample rate of audio signal
+        n_mfcc (int) : number of mfc coefficients to retain
+        dct_type (int) : type of DCT (discrete cosine transform) to use
+        norm (string) : norm to use
+        melkwargs (dict, optional): arguments for MelSpectrogram
+    """
+    def __init__(self, sr=16000, n_mfcc=40, dct_type=2, norm='ortho', melkwargs=None):
+
+        supported_dct_types = [2]
+        if dct_type not in supported_dct_types:
+            raise ValueError('DCT type not supported'.format(dct_type))
+        self.sr = sr
+        self.n_mfcc = n_mfcc
+        self.dct_type = dct_type
+        self.norm = norm
+        self.melkwargs = melkwargs
+        self.top_db = 80.
+        self.s2db = SpectrogramToDB("power", self.top_db)
+
+        if melkwargs is not None:
+            self.MelSpectrogram = MelSpectrogram(sr=self.sr, **melkwargs)
+        else:
+            self.MelSpectrogram = MelSpectrogram(sr=self.sr)
+
+    def dct(self, x, norm=None):
+        """
+        Discrete Cosine Transform, Type II (a.k.a. the DCT)
+        For the meaning of the parameter `norm`, see:
+        https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.fftpack.dct.html
+        :param x: the input signal
+        :param norm: the normalization, None or 'ortho'
+        :return: the DCT-II of the signal over the last dimension
+        """
+        x_shape = x.size()
+        l = x.new_tensor(x_shape[-1])
+
+        slices_base = [slice(x_shape_i) for x_shape_i in x_shape[:-1]]
+        v = torch.cat((x[slices_base + [slice(0, None, 2)]],
+                       x[slices_base + [slice(1, None, 2)]].flip([-1])),
+                      dim=-1)  # (*, l)
+        Vc = torch.rfft(v, 1, onesided=False)  # (*, l, 2)
+
+        k = -torch.arange(l, dtype=x.dtype, device=x.device) * np.pi / (2 * l)
+        for _ in range(Vc.dim() - 2):
+            k.unsqueeze_(0)
+        W_r, W_i = k.cos(), k.sin()  # (1,..., l)
+
+        # V is (*, l)
+        V = Vc.narrow(-1, 0, 1).squeeze(-1) * W_r - Vc.narrow(-1, 1, 1).squeeze(-1) * W_i
+
+        if norm == 'ortho':
+            V_ndim_0 = V.narrow(-1, 0, 1)
+            V_ndim_rest = V.narrow(-1, 1, x_shape[-1] - 1)
+
+            V_ndim_0.div_(torch.sqrt(l) * 2)
+            V_ndim_rest.div_(torch.sqrt(l / 2) * 2)
+
+        V.mul_(2.)
+
+        return V
```

Comment:
I find this implementation hard to read without comments. The `torch.cat` looks potentially expensive. Is this really faster than just constructing a DCT matrix and doing a matmul? The latter also easily allows to compute a limited number of `self.n_mfcc` components. Old implementation by me:
```python
def create_dct(dim, outdim=None, orthogonal=True):
    """
    Creates a DCT transformation matrix.
    @param dim: Dimensionality of input data to transform
    @param outdim: If given, create a matrix for the first 'outdim' DCT
        coefficients only. If omitted, create a full (squared) DCT matrix.
    @param orthogonal: If given, makes the transform orthogonal
    @return The transformation matrix, to be right-multiplied to row-wise data.
    """
    if outdim is None:
        outdim = dim
    elif outdim > dim:
        raise ValueError("DCT outdim cannot be larger than dim")
    # http://en.wikipedia.org/wiki/Discrete_cosine_transform#DCT-II
    n = np.arange(dim)
    k = np.arange(outdim)[:, np.newaxis]
    dct = np.cos(np.pi/dim * (n + 0.5) * k)
    if orthogonal:
        dct[0] *= 1.0 / np.sqrt(2)
        dct *= np.sqrt(2.0 / dim)
    return dct
```
This matrix could be computed at construction time (in torch or numpy) and then multiplied with in the forward pass.

Reply:
Yep, good idea. This actually gives about a ~10x speed up on CPU, maybe more on GPU. Integrating now.